### PR TITLE
[PoC] End to end vision exports

### PR DESCRIPTION
**Pull request by guarin**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48151)

# What does this PR do?

PoC to export models end to end with the new exporters from `transformers/exporters`. 

Main changes:
* Add `HfExporter.export_end_to_end`
* Add `return_traceable_outputs` to post process methods (by default `False`)
* Split each `post_process_xyz` method into `post_process_xyz` and `build_xyz_outputs`. All traceable code stays in the body of `post_process_xyz` which finishes with the call to the untraceable `build_xyz_outputs`. 

The motivation to expose `return_traceable_outputs` as a flag is that you can call it also from normal torch outside of an export context. This simplifies testing and any downstream applications that take either a plain torch model or an exported model as input.

The motivation to expose `build_xyz_outputs` as a public method is that you want to be able to call it on the outputs from the exported model to get the final, clean outputs which match normal torch post processing outputs. Note that depending on the task it is not necessary to call this method. For example, semantic segmentation can return final label predictions directly from the traced code. For other tasks like object detection and instance segmentation there will always be some extra post-processing required.

Example code:
```
import numpy as np
import torch
from PIL import Image
from transformers import AutoImageProcessor, Mask2FormerForUniversalSegmentation
from transformers.exporters.exporter_onnx import OnnxConfig, OnnxExporter

checkpoint = "facebook/mask2former-swin-tiny-ade-semantic"
processor = AutoImageProcessor.from_pretrained(checkpoint)
model = Mask2FormerForUniversalSegmentation.from_pretrained(checkpoint).eval()

image = Image.open("coco_sample.jpg").convert("RGB")
image_tensor = torch.from_numpy(np.array(image)).permute(2, 0, 1).contiguous()  # raw uint8 CHW
target_sizes = [(image_tensor.shape[-2], image_tensor.shape[-1])]   # output at original resolution

exporter = OnnxExporter()
config = OnnxConfig(dynamic=False, output_path="mask2former_semantic.onnx")

onnx_program = exporter.export_end_to_end(
    model,
    preprocess=processor,
    post_process=processor.post_process_semantic_segmentation,
    sample_inputs={"images": image_tensor, "target_sizes": target_sizes},
    config=config,
)

(semantic_map,) = onnx_program(images=image_tensor)
segmentation = processor.build_semantic_segmentation_outputs(semantic_map=semantic_map)[0]

# equivalent to
inputs = processor(image).to("cuda")
outputs = model(**inputs)
segmentation = processor.post_process_semantic_segmentation(outputs, target_sizes=target_sizes)
```

This is still very much WIP and not cleaned up. Docstrings are too verbose for now.

cc @molbap @IlyasMoutawwakil

### CODE REVIEW

File: src/transformers/exporters/exporter_onnx.py [hunk 1]
```diff
@@ -387,6 +387,32 @@ def patch(input, *shape, **kwargs):
     return patch
 
 
+@register_patch("onnx", "torch.Tensor.contiguous")
+def _patch_contiguous(original):
+    """Materialise via `clone` instead of a bare `contiguous()` call.
+
```
Comment (guarin) [inline src/transformers/exporters/exporter_onnx.py:393]: This was needed to make onnx export work with mask2former which uses swin transformer under the hood. Crash happened here: https://github.com/huggingface/transformers/blob/b6688043ff8e1c972c7fcac92db7942e1e41f42e/src/transformers/models/swin/modeling_swin.py#L465

Apparently caused by the `contiguous` getting folded away here: https://github.com/huggingface/transformers/blob/b6688043ff8e1c972c7fcac92db7942e1e41f42e/src/transformers/models/swin/modeling_swin.py#L396

Probably not the right solution, just wanted it to run through :)

File: src/transformers/exporters/utils.py [hunk 2]
```diff
@@ -927,3 +933,232 @@ def decompose_for_generation(
...
+# is a one-liner over it plus `bind_static_post_process_kwargs`.
+
+
+@register_patch(
+    "dynamo", "transformers.image_processing_backends.TorchvisionBackend._fuse_mean_std_and_rescale_factor"
+)
+@register_patch("onnx", "transformers.image_processing_backends.TorchvisionBackend._fuse_mean_std_and_rescale_factor")
+def _patch_fuse_mean_std_and_rescale_factor(original):
+    """Drop the `lru_cache` around the fused mean/std tensors while a preprocessing stage is traced.
+
```
Comment (guarin) [inline src/transformers/exporters/utils.py:952]: Probably also not the right solution but does the job for now

File: src/transformers/exporters/utils.py [hunk 3]
```diff
@@ -927,3 +933,232 @@ def decompose_for_generation(
...
+        if param.kind in (inspect.Parameter.POSITIONAL_OR_KEYWORD, inspect.Parameter.KEYWORD_ONLY)
+    ]
+    var_keyword = any(param.kind is inspect.Parameter.VAR_KEYWORD for param in parameters)
+    return declared, var_keyword
+
+
+def route_end_to_end_inputs(
+    input_names: list[str], preprocess: TorchvisionBackend | Callable | None, post_process: Callable | None
+) -> tuple[list[str], list[str]]:
+    """Split `input_names` into the names `preprocess` consumes and the names `post_process` consumes.
```
Comment (guarin) [inline src/transformers/exporters/utils.py:995]: This might also have to be reworked. Right now it magically splits arguments into preprocess, forward, post-process but might be better to be explicit about where things should end up.

File: src/transformers/exporters/exporter_onnx.py (hunk 1 above)
Comment (molbap) [reply src/transformers/exporters/exporter_onnx.py:393]: nice! I would've bet on more clone vs contiguous clashes, let's see how we can fix

File: src/transformers/exporters/utils.py [hunk 4]
```diff
@@ -927,3 +933,232 @@ def decompose_for_generation(
...
+        bound["target_sizes"] = target_sizes
+
+    if bound:
+        post_process = functools.partial(post_process, **bound)
+    return post_process, sample_inputs
+
+
+if is_torch_available():
+
+    class EndToEndModel(torch.nn.Module):
```
Comment (guarin) [inline src/transformers/exporters/utils.py:1085]: prob best to move out of this file

File: src/transformers/models/mask2former/modular_mask2former.py [hunk 5]
```diff
@@ -63,52 +68,98 @@ def post_process_semantic_segmentation(
...
+            When `return_traceable_outputs=True`, a tuple `(semantic_map,)` of shape `(batch_size, height, width)`,
+            extended with the segmentation scores of shape `(batch_size, num_classes, height, width)` if
+            `return_segmentation_scores=True`. Everything up to that tuple is traceable, so passing it to
+            [`~Mask2FormerImageProcessor.build_semantic_segmentation_outputs`] outside of the traced code gives the
+            same output as `return_traceable_outputs=False`.
         """
+        if target_sizes is not None:
+            if isinstance(target_sizes, (torch.Tensor, np.ndarray)):
+                target_sizes = target_sizes.tolist()
+            target_sizes = [tuple(size) for size in target_sizes]
```
Comment (guarin) [inline src/transformers/models/mask2former/modular_mask2former.py:82]: side note: target_sizes handling is broken in most processors. lots of cases can end up raising errors.

File: src/transformers/exporters/base.py [hunk 6]
```diff
@@ -203,3 +205,80 @@ def export_for_generation(
                 ) from e
 
         return exported
+
+    @requires(backends=("torch",))
```
Comment (IlyasMoutawwakil) [inline src/transformers/exporters/base.py:209]: no need i guess since we validate_environment in init

File: src/transformers/models/mask2former/modular_mask2former.py (hunk 5 above)
Comment (molbap) [reply src/transformers/models/mask2former/modular_mask2former.py:82]: yes.. I think we can likely abstract away all these checks with one big checker/validator util from `image_processing_utils.py` say, that would validate target sizes and convert tensor-like inputs to nested tuples of integers

File: src/transformers/exporters/utils.py [hunk 7]
```diff
@@ -927,3 +933,232 @@ def decompose_for_generation(
     components = decompose_multimodal(prefill_model, prefill_inputs)
     components["decode"] = stages["decode"]
     return components
+
+
+# ── End-to-end composition ────────────────────────────────────────────────────
+# Chain preprocessing, the model forward and post-processing into a single `nn.Module` so
+# a backend's `export` traces all three stages into one graph. `EndToEndModel` owns the
+# plumbing — routing `sample_inputs` to the stage that declares them, and aligning the
+# preprocess output with the model's dtype/device — and `HfExporter.export_end_to_end`
```
Comment (molbap) [inline src/transformers/exporters/utils.py:942]: nice, this should be in docs/corresponding .md linked documentation

File: src/transformers/models/mask2former/modular_mask2former.py [hunk 8]
```diff
@@ -63,54 +68,100 @@ def post_process_semantic_segmentation(
...
+        return self.build_semantic_segmentation_outputs(
+            segmentation_scores=segmentation_scores,
+            target_sizes=target_sizes,
+            return_segmentation_scores=return_segmentation_scores,
+        )
+
+    def build_semantic_segmentation_outputs(
+        self,
+        segmentation_scores: torch.Tensor,
+        target_sizes: list[tuple[int, int]] | None = None,
```
Comment (guarin) [inline src/transformers/models/mask2former/modular_mask2former.py:126]: @molbap as discussed offline: it would be nice to have different backends (numpy/torch) for these `build` methods as depending on where you run the exported model the tensors might be from a different library. E.g. onnx with numpy as input/output. But it could also be C++ or some hardware specific code. So those are actually the "true" post-processing methods while the body of `post_process_xyz` belongs more to the model code and we can assume it runs in torch.

Imo not super important to address this right now but something to keep in mind. Overall the split of the post-processing methods is a bit awkward, especially to keep everything BC compatible.

File: src/transformers/exporters/utils.py [hunk 9]
```diff
@@ -927,3 +933,232 @@ def decompose_for_generation(
...
+@register_patch(
+    "dynamo", "transformers.image_processing_backends.TorchvisionBackend._fuse_mean_std_and_rescale_factor"
+)
+@register_patch("onnx", "transformers.image_processing_backends.TorchvisionBackend._fuse_mean_std_and_rescale_factor")
+def _patch_fuse_mean_std_and_rescale_factor(original):
+    """Drop the `lru_cache` around the fused mean/std tensors while a preprocessing stage is traced.
+
+    The cache is keyed on the processor and its normalization kwargs, so the fake tensors the first
+    trace puts in it are handed back to every later call — poisoning the next export ("we found a
+    fake tensor in the exported program constant's list") and eager preprocessing alike. Tracing
```
Comment (molbap) [inline src/transformers/exporters/utils.py:955]: what is in quotes here exactly?

File: src/transformers/exporters/base.py (hunk 6 above)
Comment (molbap) [reply src/transformers/exporters/base.py:209]: I think `requires` goes first, IIRC, we need it for lazy init no?

File: src/transformers/exporters/utils.py [hunk 10]
```diff
@@ -927,3 +933,232 @@ def decompose_for_generation(
...
+    """
+    sample_inputs = dict(sample_inputs)
+    if post_process is None:
+        return post_process, sample_inputs
+
+    declared, _ = _stage_input_names(post_process, skip_first=True)
+    bound: dict[str, Any] = {}
+
+    if "return_traceable_outputs" in declared:
+        bound["return_traceable_outputs"] = True
```
Comment (molbap) [inline src/transformers/exporters/utils.py:1067]: I think this is the biggest issue here (overall I like the design, but the split makes it awkward). I htink if the torch part had its own method `post_process_*_static` followed by the python/numpy tail, it'd be clearer, with the `post_process_*` being this "exportable body" followed by its non-exportable tail. that way we can drop this flag. 

It'd add a level of granularity conditioned on "is this torch or not" but I think it could be acceptable

File: src/transformers/exporters/utils.py (hunk 4 above)
Comment (molbap) [reply src/transformers/exporters/utils.py:1085]: yes, likely better in its own file?

File: src/transformers/models/mask2former/modular_mask2former.py [hunk 11]
```diff
@@ -164,71 +228,127 @@ def post_process_instance_segmentation(
...
+        scores, topk_indices = scores.flatten(1, 2).topk(num_queries, dim=-1, sorted=False)
+        classes = topk_indices % num_classes
+        query_indices = torch.div(topk_indices, num_classes, rounding_mode="floor")
 
-        # Loop over items in batch size
-        results: list[dict[str, TensorType]] = []
+        mask_logits = masks_queries_logits.gather(
+            dim=1, index=query_indices[..., None, None].expand(-1, -1, height, width)
+        )
+        masks = (mask_logits > 0).to(mask_logits.dtype)
```
Comment (molbap) [inline src/transformers/models/mask2former/modular_mask2former.py:262]: eh, when running this in `float16` it failed here. Not sure if it matters (mask comes up as `None`)

**Review by molbap — COMMENTED** [review]
Finally had time to play around with the exporter. I think if we add a method for the "static" part we can keep it in the processor rather than in the modeling. left a couple other comments
